crates/rust-analyzer/src/reload.rs

```diff
@@ -296,11 +296,25 @@ impl GlobalState {
         let workspaces =
             workspaces.iter().filter_map(|res| res.as_ref().ok().cloned()).collect::<Vec<_>>();
 
-        let same_workspaces = workspaces.len() == self.workspaces.len()
-            && workspaces
-                .iter()
-                .zip(self.workspaces.iter())
-                .all(|(l, r)| l.eq_ignore_build_data(r));
+        // `different_workspaces` is used to determine whether to spawn a a new proc macro server for
+        // a newly-added rust workspace (most commonly sourced from a `rust-project.json`). While the
+        // algorithm to find the new workspaces is quadratic, we generally expect that the number of total
+        // workspaces to remain in the low single digits. the `cloned_workspace` is needed for borrowck
+        // reasons.
+        let cloned_workspaces = workspaces.clone();
+        let different_workspaces = cloned_workspaces
```

Comment:
I can undo these changes and go back to the previous approach, since the reason I even did this proved to be wrong 😅

Reply:
We don't even need the checks for whether the new workspaces are rust-project.json ones or not, we need respawn the proc-macro servers in general when `!same_workspaces`. (Think of when someone adds a new top level cargo project)

Reply:
Oh, good point. Didn't consider that; will add!